internal/core/command/device.go

```diff
@@ -94,24 +127,34 @@ func executeCommandByName(
 		return "", err
 	}
 
-	return executeCommandByDevice(ctx, d, command, body, queryParams, isPutCommand, lc)
+	return executeCommandByDevice(ctx, d, command, body, lc, r, httpCaller)
 }
 
 func executeCommandByDevice(
 	ctx context.Context,
 	device contract.Device,
 	command contract.Command,
 	body string,
-	queryParams string,
-	isPutCommand bool,
-	lc logger.LoggingClient) (string, error) {
+	lc logger.LoggingClient,
+	r *http.Request,
+	httpCaller internal.HttpCaller) (string, error) {
 
+	var method string
 	var ex Executor
 	var err error
-	if isPutCommand {
-		ex, err = NewPutCommand(device, command, body, ctx, &http.Client{}, lc)
+
+	if r == nil && r.Method == "" {
+		return "", errors.NewErrRequestParameterSupplied(r.Method)
 	} else {
-		ex, err = NewGetCommand(device, command, queryParams, ctx, &http.Client{}, lc)
+		method = r.Method
+		switch method {
+		case http.MethodPut:
+			ex, err = NewPutCommand(device, command, body, ctx, httpCaller, lc, r)
+		case http.MethodGet:
+			ex, err = NewGetCommand(device, command, ctx, httpCaller, lc, r)
+		default:
+			lc.Error(fmt.Sprintf("unknown method: %s", method))
+		}
```

Comment:
See my other comment regarding inverting the logic here for simplicity, and elimination of side effects. Also, there seems to be a bug here with the `error` in the `if` portion. The `error` uses `r.Method` directly after ensuring that it is blank, which means it is always empty.